For #54: new architecture proposal

paulodamaso opened the pull request:
For #54
New architecture draft:
Added package `com.artipie.maven.artifact`, with artifact abstractions:
- `Artifact`: maven artifact abstraction
- `Coordinates`: artifact coordinates (groupId, artifactId, version)
- `Metadata`: artifact metadata (path, size, md5, sha)
- `Downloaded`: downloaded artifact using `Publisher<ByteBuffer>`
- `Uploaded`: uploaded maven artifact
Added `package com.artipie.maven.repository`, with repository abstractions:
- `Repository`: maven respository
- `Proxied`: maven repository with proxy

This is a very preliminar draft for architecture change, but the idea is to build the project around `Repository` implementations / decorations to get the maven artifacts.

paulodamaso commented:
@g4s8 Please take a look at this draft

g4s8 commented:
@paulodamaso I like the design, but I didn't find one important thing: there're no methods to update metadata for existing artifact. Let me show you one common scenario of usage this library:
 1. user already have a set of artifacts, e.g, `com.artipie:asto:1.0`, `com.artipie:asto:2.0`, `com.artipie:asto:3.0` and valid metadata file for these artifacts, see this link for example: https://repo.maven.apache.org/maven2/com/artipie/asto/
 2. user adds new version `0.4` for this artifact and wants to **update** metadata files
 3. user uses maven-adapter library to do that, user uses `Storage` interface from `artipie/asto` library to work with artifact files, user calls `new Maven(storage).update(new Key.From("com/artipie/asto"));` to regenerate the metadata

This is one of the primary use-cases, how we can put it into your design?

paulodamaso commented:
@g4s8 We can:
- do it automatically on Artifact upload, e.g. `new Uploaded(new Artifact()).value()`
- do it on `Artifact (new Artifact().update())`
WDYT?

0crat commented:
This pull request [#55](https://github.com/artipie/maven-adapter/issues/55) is assigned to @g4s8[/z](https://www.0crat.com/u/g4s8), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/023de864-3b6d-4235-a9d4-b908262d060c); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

<!-- https://www.0crat.com/footprint/CT2E6TK9B/386ed3e6-e0bc-4e3e-9a91-540e65809a63, version: 0.51.9, hash: ${buildNumber} -->

g4s8 commented:
@paulodamaso these options won't work since they don't encapsulate `Storage` interfaces. I'd start with `new Maven(storage).update(new Coordinates("com.artipie:asto:0.1"))` to keep interfaces of all adapters similar. Let me show you the plan:
 - first we need to implement metadata generators till 1 April 2020
 - the next step is to implement `Slice` interface from `artipie/http` with API for Maven repository till 1 May 2020.

We actually don't need uploading methods for Maven, since it's already implemented in `Storage` interface. The main purpose of this adapter is to generate metadata for existing artifacts in storage.

paulodamaso commented:
@g4s8 Sorry I didn't put all the needed details into my explanation. In the proposed architecture 
`Artifact` and `Repository` implementation would both encapsulate `Storage`. In this case we would have artifact regeneration declaratively like this:
`new Metadata(new Artifact(storage))`
How about?

paulodamaso commented:
@g4s8 Sorry for the delay, added some puzzles to start by, and deprecated all outdated libraries.

g4s8 commented:
@paulodamaso thanks! Just a few points:
 - we don't need upload and download objects, since it will be managed by `Storage` itself
 - the main purpose of this adapter is to generate metadata for existing maven repo files, so please don't introduce interfaces or objects "for future". We'll discuss next steps later after first release. So I'd prefer to reduce the scope of this adapter to the only one thing: generating metadata for files in `Storage` by `Key` and put this metadata to the same `Storage`. E.g. if we have an artifact in storage at key `com/artipie/maven-adapter/0.1`, we should be able to run something like `update("com/artipie/maven-adapter/")` and this adapter will update the file at key `com/artipie/maven-adapter/metadata.xml`
 - We need to support update by `Key` interface from `artipie/asto`, since it's the core object of all system. To keep the architecture of all adapters similar, I'd create interface or class like `Maven` or `MavenRepo` with method `CompletableFuture<Void> Maven.update(Key key)`, what do you think?

paulodamaso commented:
@g4s8 Thanks for the info, things are a lot clear now!
I've solved some of the issues:
- removed all download and upload routines from new architecture;
- added key to `Artifact`, so now we depends on `Key` interface from  `artipie/asto`
- I still didn't got what `update` means. It means that we are generating new metadata for the artifact? I'm modeling this in a way that we don't need to keep running `update` for our objects, for example:
```
  new Artifact(
    storage,
    key
).metadata()
```
Where we will calculate the metadata values as needed from the `Artifact` that we have from inside the storagem, with the key. Does it makes sense?

g4s8 commented:
@paulodamaso sorry, I didn't say it correctly. The adapter needs to **update** metadata, not only generate it. This means that we may use this adapter for updating the metadata for existing Maven repository files. Let's say we have an Artifactory Maven repository, whici is stored on a file system or S3 storage. We should be able to swtich to Artipie and keep publishing to existing repository. E.g. If we had this file structure:
```tree
com/artipie/http
├── maven-metadata.xml
├── maven-metadata.xml.md5
├── maven-metadata.xml.sha1
├── 0.1
│    ├── http-01.jar
│    ├── http-01.jar.asc
│    ├── http-01.jar.md5
│    ├── http-01.jar.sha1
│    ├── http-0.1-source.jar
│    ├── http-0.1-source.jar.md5
│    ├── http-0.1-source.jar.asc
│    ├── http-0.1-source.jar.sha1
│    ├── http-0.1-javadoc.jar
│    ├── http-0.1-javadoc.jar.asc
│    ├── http-0.1-javadoc.jar.md5
│    └── http-0.1-javadoc.jar.sha1
└── 0.2 // same files
```
and we wants to add new version `0.3` and **update** existing metadata at `maven-metadata.xml` (and `.sha1` with `.md5` hashsum files). So we need some kind of entry point for this library, usually it's named as a repository type (like `Maven`, `Rpm`, `Npm`, etc) with `update(Key)` method which **updates** the metadata for artifact.
Is it clearer now?

paulodamaso commented:
@g4s8 Now I got it, thanks. It wasn't clear if updating and adding metadata were the same operation. There are still some doubts / concerns:
- I couldn't figure exactly what `Metadata` should be. At first we may think that it is `ArtifactMetadata`, but `Artifact` is defined by a `coordinate` which is composed by `groupId`, `artifactId` and `version`. So, if I try to read the metadata for `com.artipie:http:0.1` the result should be equal to the metadata from `com.artipie:http:0.3` (as the version is ignored for the metadata, as explained in the repository structure above). Is this relation between metadata and artifact correct?
- if we want to add new versions to the repository using the adapter, we should allow the `upload` routine. So, we will need  `Uploaded`  class to handle with this: save the artifacts to asto `Storage`;
- I propose that we try to do this metadata update declaratively, something like this:
```
//declares the metadata
final Metadata metadata =
new Artifact(
  storage,
  "com/artipie/http" //key
).metadata();

//saves the metadata value to maven-metadata.xml
new Updated(metadata).value()
```
where:
- `Artifact.metadata()` calculates the metadata for some specific `Artifact` (note that we are passing 
`Storage` and artifact `Key`; we can pass `Repository` instead of `Storage` if better); we will use lazy loading to access the metadata on repository only when we will use it; but I am still not certain about `metadata` method; since metadata is calculated maybe we should have the following: `new Metadata(new Artifact(storage, "com/artipie/http"))` instead the code presented above. What would be better, in your opinion?
- `Updated` updates the `maven-metadata.xml`, which can be composed with `Checksum` implementations to save each checksum to repository;

paulodamaso commented:
@g4s8 Made another commit based on some research that I've made on maven docs. Maven is a lot confused about these concepts because they mix dependency resolution and artifact storage concepts into the same implementation. Looks like things are a little messed up, but I managed to come with a decent design. 
Basically:
- the upmost element in this architecture is the `Repository`. 
- we have libraries inside the repository. Libraries are defined by `groupId`+`artifactId`, and are collections of `Artifact`, which are sorted by their `Coordinates` (`groupId`+`artifactId`+`version`)
- each artifact is composed of a unlimited number of  `File`s. These files are stored somewhere (we're dealing with just interfaces here)
- we calculate the checksums of  each `File`
Notice that we are not using any `Storage` or `Key` references because we are just talkiing about interfaces here; I'm creating puzzles to implement each element using artipie/asto Storage.

Does it makes any sense?

g4s8 commented:
@paulodamaso I agree that we may want to upload artifact using adapter, but we'll handle this later, right now it only increases the complexity of solution, but we need something like MVP with minimum functionality at first release, this is why I'd keep as few interfaces as possible. So now we don't need to extend functionality, we only need to refactor existing solution.

I like this idea `Artifact.metadata()`, but it's not fully correct in scope of the requirements: we don't have artifact metadata, only package metadata (which includes all versions of package = artifacts). Also, by the requirements we don't need to read and parse it as a part of public API, public API of this adapter includes only updating metadata file of the package. So please keep as few public interfaces as possible, it'll increase the scope, but we don't want it right now.

Also, the client code works with `Storage` and `Key` concepts, not with `Artifact`, `Repository`, `Metadata` etc, e.g. we don't have a place where `Artifact` interface can be used.

What about `new Updated(metadata).value()` I don't really like this class, because it looks like a function, not an object. The behavior here is `Updated`, not the `value`.

So I'd go with `Maven.update(Key artifact)` to be convenient with all other adapters, since all of them will be used in one place. Most probably we'll need to implement `batchUpdate(Key package)` after that with transactional of all artifacts in the package, so `Maven` object may be a good place to keep all public API of adapter.

paulodamaso commented:
@g4s8 made a new proposal, based on all the discussion and studyof the other adapters:
- looking at the other adapters I understood about `Maven`  entrypoint class. I've made a bare sketch of this
- removed most of the interfaces, retaining only `Metadata`, `Artifact` and `File`. `Artifact` represents a maven artifact which holds the `Files`. We access them using `Key` and `Storage`, as requested. `Metadata` wraps an `Artifact` and builds it metadata information.
- turned `Checksum` into an abstract class with little behavior. Now we just implement a new `Checksum` if we want to use a new type of algotythm.

I think that it as simpler as it gets, with the minimum features required to read and update metadata.

WDYT?

g4s8 commented on src/main/java/com/artipie/maven/Maven.java:
```diff
@@ -0,0 +1,63 @@
+     * Updates the metadata of a maven package.
+     * @param artifact Asto key of maven artifact.
+     */
+    public void update(final Key artifact) {
```
@paulodamaso we need to return `CompletionStage` here

paulodamaso replied:
@g4s8 Done!

g4s8 commented on src/main/java/com/artipie/maven/file/File.java:
```diff
@@ -0,0 +1,53 @@
+ *  Files must be stored somewhere. Use artipie/asto to provide an storage for
+ *  the repository files.
+ */
+public interface File {
```
@paulodamaso what's the reason of this class? Can't we use just `java.nio.Path` here?

paulodamaso replied:
@g4s8 I intended to abstract the maven file behavior, allowing us to store maven files in any storage media we want, not just filesystem. That would allow us to use an `AbstractStorage` for this, too.

g4s8 replied:
@paulodamaso so it's only for temporary files, correct? Don't forget that we're using `Publisher<ByteBuffer>` for data in `Storage`: `Publisher<ByteBuffer> data = storage.value(key)`

paulodamaso replied:
@g4s8 Actually they are the actual maven repository files: these files can be stored in any kind of media (filesystem, database, asto storage), so I thought in abstracting its behavior.

g4s8 commented on src/main/java/com/artipie/maven/metadata/Metadata.java:
```diff
@@ -0,0 +1,80 @@
+        }
+
+        @Override
+        public Flow.Publisher<ByteBuffer> content() {
```
@paulodamaso it should be `Publisher<ByteBuffer>` from `org.reactivestreams` dependency (the only difference is that `Flow.Publisher` is standard JDK9 interface, but `org.reactivestreams` interfaces are from external dependencies and works on JDK8 too. We support JDK8 for adapters for library clients, but using JDK11 for server).

paulodamaso replied:
@g4s8 Done!

paulodamaso commented:
@g4s8 Corrections done and question answered, please take a look

g4s8 commented:
@paulodamaso thanks. Can you fix the build and we can merge?

paulodamaso commented:
@g4s8 I got the following problems:
- when using java 8, asto libreris arent's compatible;
- when using java 11, qulice libraries aren't compatibles.

How to proceed?

g4s8 commented:
@paulodamaso asto is built with `1.8` target SDK, please use `com.artipie:ppom:0.3` as parent pom, see https://github.com/artipie/rpm-adapter/blob/master/pom.xml

paulodamaso commented:
@g4s8 We have a lot of code in java 10 in deprecated classes which prevents the project to compile, should I get rid of it?

g4s8 commented:
@paulodamaso yes, but you can do it later (todo puzzle would be enough), all artipie adapters should be 1.8 compatible, but the server and http layer can use higher version of JDK

paulodamaso commented:
@g4s8 I'm having compilation problems in the file, there are a lot of `var` references. At the same time, I don't want to remove deprecated code because we might use some of them in our implementations.

paulodamaso commented:
@g4s8 PLease take a look now, build is clean. Have excluded all previous code from compile and tests since we won't be using it anymore, but we can extract pieces of it to implement some routines.

0crat commented:
Code review was too long (12 days), architects (@g4s8) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

<!-- https://www.0crat.com/footprint/CT2E6TK9B/a926f4b8-56d9-4e35-b876-e50040df2b81, version: 0.51.12, hash: ${buildNumber} -->
